container_images/pytest/main.py

```diff
@@ -0,0 +1,166 @@
+      containing directory will be recursively copied to the execution root.
+
+  Returns:
+    Absolute path to execution root.
+  """
+  assert (os.path.isabs(pyproject_toml_file) and
+          os.path.exists(pyproject_toml_file))
+
+  source_root = os.path.dirname(pyproject_toml_file)
+  exec_root = tempfile.mkdtemp()
+  shutil.copytree(source_root, exec_root, dirs_exist_ok=True)
+  return exec_root
+
+
+def to_tox_version(py_version: str):
+  """Convert `{major}.{minor}` to `py{major}{minor}`"""
+  if re.fullmatch(r"\d.\d{1,2}", py_version):
+    major, minor = py_version.split(".")
+    return "py{major}{minor}".format(major=major, minor=minor)
+  else:
+    sys.exit("Invalid version number: " + py_version)
+
+
+def write_tox_ini():
+  """Read project configuration from pyproject.toml, and write a new tox.ini
+  file.
+
+  The tox.ini file is generated to ensure that tests are run consistently,
+  and that test reports are written to the correct location.
+  """
+  assert os.path.exists("pyproject.toml")
+
+  with open("pyproject.toml") as f:
+    project_cfg = toml.load(f)
+    test_cfg = project_cfg.get("tool", {}).get("gcp-guest-pytest", {})
+
+  # Which interpreters to enable.
+  envlist = [to_tox_version(env) for env in test_cfg.get("envlist", [])]
+
+  # Test-specific dependencies to install.
+  test_deps = test_cfg.get("test-deps", [])
+
+  if not envlist:
+    sys.exit("pyproject.toml must contain a section [tool.gcp-guest-pytest] "
+             'with a key "envlist" and at least one interpreter.')
+
+  config = configparser.ConfigParser()
+  config["tox"] = {
+    "envlist": ", ".join(envlist),
+  }
+
+  config["testenv"] = {
+    "envlogdir": "{env:ARTIFACTS}/tox/{envname}",
+    "deps": "\n\t".join(_common_test_deps + test_deps),
+    "commands": " ".join(_per_interpreter_command),
+  }
+
+  if os.path.exists("tox.ini"):
+    print("Ignoring existing tox.ini file.")
+    os.remove("tox.ini")
+
+  with open("tox.ini", "w") as f:
+    config.write(f)
+
+
+def save_configs(dst: str):
+  print("Saving config files to " + dst)
+  if not os.path.exists(dst):
+    os.mkdir(dst)
+  for fname in ["pyproject.toml", "tox.ini"]:
+    shutil.copy2(fname, dst)
+
+
+def validate_args(args: typing.List[str]):
+  if (len(args) > 1 and args[1].endswith("pyproject.toml") and
+      os.path.exists(args[1])):
+    toml_file = os.path.abspath(args[1])
+  else:
+    sys.exit("First argument must be path to pyproject.toml")
+  if "ARTIFACTS" in os.environ and os.path.exists(os.environ["ARTIFACTS"]):
+    artifact_dir = os.path.abspath(os.environ["ARTIFACTS"])
+    # Protect against a relative artifacts directory.
+    os.environ["ARTIFACTS"] = artifact_dir
+  else:
+    sys.exit("$ARTIFACTS must be set to a directory that exists")
+  return artifact_dir, toml_file
+
+
+def main():
+  print("args: %s" % sys.argv)
+  print("$ARTIFACTS: %s" % os.environ.get("ARTIFACTS"))
+  artifact_dir, toml_file = validate_args(sys.argv)
```

Comment:
ah, if we are taking the config in as an argument, that means the prow job will have to be configured with the path. that's fine, but just to be clear that's part of the design here and we aren't assuming any location or doing any searching for it

Reply:
Ahh, I see, we're using directory paths in the other prow jobs. Switched to a directory rather than the config file.